src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,545 @@
+    require(_pendingProposals.add(hash), ErrExistedProposal(hash));
+
+    emit ProposalCreated(msg.sender, p.proposer, hash, p, $._createdAt);
+  }
+
+  /**
+   * @dev Snapshot the vote power of all candidates at the time of proposal creation.
+   */
+  function _snapshotVotePower(
+    bytes32 hash
+  ) internal returns (uint256 totalVotePower) {
+    address[] memory allCids = ICandidateManager(getContract(ContractType.VALIDATOR)).getValidatorCandidateIds();
+    uint256[] memory selfStakes = IStaking(getContract(ContractType.STAKING)).getManySelfStakingsById(allCids);
+
+    uint256 length = allCids.length;
+    require(length == selfStakes.length, ErrLengthMismatch(msg.sig));
+
+    ProposalInfo storage $ = _info[hash];
+
+    for (uint256 i; i < length; ++i) {
+      uint96 scaledStake = (selfStakes[i] / 1 ether).toUint96();
+      totalVotePower += scaledStake;
+
+      $._votePower[allCids[i]] = scaledStake;
+
+      // Self-stake is expected to > 1e18
+      require(scaledStake != 0, ErrPanicDataCorruption(allCids[i]));
+    }
+  }
+
+  /**
+   * @dev Vote on a proposal.
+   */
+  function _vote(bytes32 hash, address cid, Vote v) internal {
+    ProposalInfo storage $ = _info[hash];
+    IProfile profile = IProfile(getContract(ContractType.PROFILE));
+
+    _requireValidNonce($._proposal.nonce);
+    require(v == Vote.For || v == Vote.Against, ErrInvalidVote());
+    require($._votePower[cid] != 0, ErrNewlyRegisteredOrRenouncedOrEmergencyExit());
+    require($._vote[cid] == Vote.Unknown, ErrAlreadyVoted(hash, cid, $._vote[cid]));
+    require(
+      _computePeriod(profile.getId2RegisteredAt(cid) + COOLDOWN_PERIOD) <= _computePeriod($._createdAt),
+      ErrNewlyRegisteredCannotVote(cid)
+    );
+
+    $._vote[cid] = v;
+    if (v == Vote.For) {
+      $._votePow.vFor += $._votePower[cid];
+    } else {
+      $._votePow.vAgainst += $._votePower[cid];
+    }
+
+    emit Voted(msg.sender, cid, hash, v, $._votePower[cid]);
+  }
+
+  /**
+   * @dev Try to execute a proposal.
+   *
+   * - Skip if the proposal is not passed.
+   * - Skip if the proposal is not executed by the contract itself or `msg.sender`.
+   * - Skip if the proposal is cancelled.
+   * - Remove the proposal from the pending set if transitioned to `Passed` or `Cancelled` or `Executed`.
+   *
+   * Emits a {ProposalCancelled} event if the proposal is cancelled.
+   * Emits a {ProposalPassed} event if the proposal is passed (current caller is not executor).
+   * Emits a {ProposalExecuted} event if the proposal is executed
+   *
+   * @return executed true if the proposal is executed.
+   */
+  function _tryExecute(
+    bytes32 hash
+  ) internal returns (bool executed) {
+    ProposalInfo storage $ = _info[hash];
+
+    if ($._votePow.vAgainst > _calcMinVoteAgainst($._totalVotePow)) {
+      $._state = State.Cancelled;
+      emit ProposalCancelled(msg.sender, hash);
+      return false;
+    }
+
+    if ($._votePow.vFor < _calcMinVoteFor($._totalVotePow)) return false;
```

Comment:
Note that in the case where `totalVotePow = 101` and the threshold is `70%`, if `$._votePow.For = 70`, it would not satisfy `>= 101 * 70% = 70.7`. To be precise, I believe the comparison should be whether `$._votePow.vFor * denom >= totalVotePow * num`. Similarly, for early proposal cancellation, we need to compare `$._votePow.Against * denom > totalPow * (denom - num)`.